R/r-session.R

```diff
@@ -0,0 +1,306 @@
+
+#' External R Session
+#'
+#' A permanent R session that runs in the background. This is an R6 class
+#' that extends the [processx::process] class.
+#'
+#' The process is started at the creation of the object, and then it can
+#' be used to evaluate R function calls, one at a time.
+#'
+#' @section Usage:
+#' ```
+#' rs <- r_session$new(options)
+#' rs$run(func, args = list())
+#' rs$call(func, args = list())
+#' rs$wait_for_call(timeout = -1)
+#' rs$get_result()
+#' rs$get_running_time()
+#' rs$get_state()
+#' rs$finish()
+#' ```
+#'
+#' @section Arguments:
+#' * `options`: A list of options created via [r_session_options()].
+#' * `func`: Function object to call in the background R process.
+#' *   Please read the notes for the similar argument of [r()]
+#' * `args`: Arguments to pass to the function. Must be a list.
+#' * `timeout`: Timeout in milliseconds.
+#'
+#' @section Details:
+#' `r_session$new()` creates a new R background process. It returns
+#' immediately, i.e. before the process is actually ready to run. You may
+#' call `wait_for_call()` to make sure it is ready.
+#'
+#' `rs$run()` is similar to [r()], but runs the function in the `rs` R
+#' session. Note that if a timeout happens, the session and the background
+#' computation is not terminated. You can call `rs$finish()` to terminate
+#' the R process. There is currently no way to terminate the computation
+#' without terminating the background R process.
+#'
+#' `rs$call()` starts running a function in the background R session, and
+#' returns immediately. To check if the function is done, call the
+#' `wait_for_call()` method. To get the result call the `get_result()`
+#' method.
+#'
+#' `rs$wait_for_call()` waits for an `rs$call()` computation, or the R
+#' session startup to finish. This is essentially a poll operation.
+#' If there is no computation running, it returns immediately.
+#'
+#' `rs$get_result()` returns the result of the last `rs$call()`
+#' computation. (Or the result of the last `rs$run()`, if it was
+#' interrupted.) If there is no result to return, because the computation
+#' has not finished yet, or some other reason, it throws an error.
+#'
+#' `rs$get_running_time()` returns the elapsed time since the R process
+#' has started, and the elapsed time since the current computation has
+#' started. The latter is NA if there is no active computation.
+#'
+#' `rs$get_state()` return the state of the R session. Possible values:
```

Comment:
Do we need a way to get the status code if the process has finished?

Reply:
It  inherits all  methods from `processx::process`, so you have `rs$get_exit_status()`.